File: apis/src/main/java/com/beat/apis/member/facade/MemberFacade.java
package com.beat.apis.member.facade;

import org.springframework.stereotype.Service;

import com.beat.apis.member.application.AuthenticationService;
import com.beat.apis.member.application.SocialLoginService;
import com.beat.apis.member.application.dto.request.MemberLoginRequest;
import com.beat.apis.member.application.dto.response.AccessTokenGenerateResponse;
import com.beat.apis.member.application.dto.response.LoginSuccessResponse;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class MemberFacade {
	private final AuthenticationService authenticationService;
	private final SocialLoginService socialLoginService;

	public LoginSuccessResponse handleSocialLogin(String authorizationCode, MemberLoginRequest loginRequest) {
		return socialLoginService.handleSocialLogin(authorizationCode, loginRequest);
	}

	public AccessTokenGenerateResponse generateAccessTokenFromRefreshToken(String refreshToken) {
		return authenticationService.generateAccessTokenFromRefreshToken(refreshToken);
	}

	public void signOut(Long memberId) {
		authenticationService.signOut(memberId);
	}
}

File: apis/src/main/kotlin/com/beat/apis/member/facade/MemberFacade.kt
package com.beat.apis.member.facade

import com.beat.apis.member.api.request.MemberLoginRequest
import com.beat.apis.member.api.type.SocialTypeRequest
import com.beat.apis.member.api.response.AccessTokenGenerateResponse
import com.beat.apis.member.api.response.MemberLoginResponse
import com.beat.application.frontoffice.auth.command.AuthenticationCommandService
import com.beat.application.frontoffice.member.command.SocialLoginCommand
import com.beat.application.frontoffice.member.command.SocialLoginCommandService
import com.beat.application.frontoffice.member.command.SocialLoginType
import org.springframework.stereotype.Service

@Service
class MemberFacade(
    private val authenticationCommandService: AuthenticationCommandService,
    private val socialLoginCommandService: SocialLoginCommandService,
) {
    fun handleSocialLogin(authorizationCode: String, request: MemberLoginRequest): MemberLoginSession =
        socialLoginCommandService.handleSocialLogin(
            authorizationCode,
            SocialLoginCommand(
                socialType = request.socialType.toApplicationType(),
            ),
        ).let { result ->
            MemberLoginSession(
                response = MemberLoginResponse.from(result),
                refreshToken = result.refreshToken,
            )
        }

    fun generateAccessTokenFromRefreshToken(refreshToken: String): AccessTokenGenerateResponse =
        AccessTokenGenerateResponse.from(authenticationCommandService.generateAccessTokenFromRefreshToken(refreshToken))

    fun signOut(memberId: Long) = authenticationCommandService.signOut(memberId)

    private fun SocialTypeRequest.toApplicationType(): SocialLoginType = when (this) {
        SocialTypeRequest.KAKAO -> SocialLoginType.KAKAO
    }
}

data class MemberLoginSession(
    val response: MemberLoginResponse,
    val refreshToken: String,
)
